[mono] Use more idiomatic cmake style

lambdageek opened the pull request:
Make mono's build use a slightly more idiomatic `cmake` style:
* One `project()` per subdirectory, rather than just in `mini`
* toplevel `add_subdirectory` for each subdirectory
* specify sources relative to the current project source dir, rather than using `addprefix` to relativize everything from `mini`
* use object libraries for each logical grouping.  Set compile definitions, link libraries, etc on the object libraries.  
* Have the final toplevel exes and shared libs (in mini; and components - when components are built to shared libs) pull in the object libraries. When building a static libmonosgen-2.0.a, pull in all the obejct libraries' objects to make a complete archive.

Doesn't untangle `src/mono/mono/components`  from `src/mono/moni/mini` yet, although things could be grouped better there too.

lambdageek committed:
make eglib into a cmake object library

The goal is to use the cmake targets instead of _sources and _headers
variables as well as to use the cmake add_subdirectory commands to
bring in parts of the build

lambdageek committed:
make mono/utils into a cmake object library

lambdageek committed:
make sgen into a cmake object library

lambdageek committed:
make metadata into a cmake object library

lambdageek committed:
set MONO_DLL_EXPORT when building the object libraries

lambdageek committed:
add eglib_api target

unlike eglib_objects, this one can be added by executables and static
or shared libraries without pulling in the eglib sources.  This is
what we want for dynamic runtime components, static interpreter and
icall table libs, etc

lambdageek committed:
when building a static library, include all the ojects

otherwise it would only include the mini objects - it's not transitive

lambdageek committed:
nit use target_compile_definitions for sgen

lambdageek committed:
log profiler links eglib objects, not just headers

am11 commented on src/mono/mono/eglib/CMakeLists.txt:
```diff
@@ -1,3 +1,4 @@
+project(eglib)
```
nit: we can also add `C` language to the relevant `project` directives. e.g. https://github.com/dotnet/runtime/blob/33e912de1027bc262ecc8ba6642359508fffc703/src/native/libs/System.IO.Compression.Native/CMakeLists.txt#L1 it makes sure that when building this project, only initialize the required toolchain instead of "all default toolchain" (in most cases under `src/mono`, C++ is not needed)

lambdageek replied:
ok, I set it.  The interesting one is `mini`. I did `project(mini C)` and then a separate `enable_language(CXX)` under some conditions.

lambdageek committed:
set project languages to C

except for mini where we also enable CXX under some conditions (LLVM
or ICU, or browser wasm)

lambdageek committed:
fix darwin framework builds

vargaz commented:
The reason we were not doing this is that metadata/ etc. dirs are not buildable by themselves, they are just used to organize source files. Previously in mono/mono, there were actually libmetadata.a files somebody can link against to build monodis, 
etc., but now we only have one libmonosgen.a file.

lambdageek commented:
> The reason we were not doing this is that metadata/ etc. dirs are not buildable by themselves, they are just used to organize source files. Previously in mono/mono, there were actually libmetadata.a files somebody can link against to build monodis, etc., but now we only have one libmonosgen.a file.

None of that changes.  there's no libmetadata.a, etc.  It's only changing how the build is organized.

lambdageek committed:
fix for older cmake

can't have sources directly on add_library(target INTERFACE ...)

lambdageek commented:
This also seems to speed up the native runtime builds (by about 2%, but still...)

methodology:

```console
killall dotnet
rm -rf artifacts
./build.sh --restore -s mono -c Debug
time ./build.sh --build -s mono -c Debug
```

baseline: 2:06.86 total
this PR: 2:03.99 total

lambdageek committed:
fix windows build

lambdageek committed:
remove uneeded define

it's set later in the cmake file already

akoeplinger reviewed (approved):
looks good to me

lambdageek commented:
/azp run runtime-extra-platforms

lambdageek commented:
runnning extra-platforms to check whether I broke the mono.mscordbi build

lambdageek commented:
mscordbi links ok. we're good
